src/services/services.ts

```diff
@@ -1829,7 +1829,35 @@ module ts {
             if (version !== sourceFile.version) {
                 // Once incremental parsing is ready, then just call into this function.
                 if (!disableIncrementalParsing) {
-                    let newSourceFile = updateSourceFile(sourceFile, scriptSnapshot.getText(0, scriptSnapshot.getLength()), textChangeRange, aggressiveChecks);
+                    let newText: string;
+                    
+                    // grab the fragment from the beginning of the original text to the beginning of the span
+                    let prefix = textChangeRange.span.start !== 0
+                        ? sourceFile.text.substr(0, textChangeRange.span.start)
+                        : "";
+                    
+                    let textChangeRangeEnd = textChangeRange.span.start + textChangeRange.span.length;
+                    // grab the fragment from the end of the span till the end of the original text
+                    let suffix = textChangeRangeEnd !== sourceFile.text.length
+                            ? sourceFile.text.substr(textChangeRangeEnd)
+                            : "";
+
+                    if (textChangeRange.newLength === 0) {
+                        // edit was a deletion - just combine prefix and suffix
+                        newText = prefix && suffix ? prefix + suffix : prefix || suffix;

```

Comment:
Cute.  But ideally unnecessary.  I would hope that the engines would optimize concats of zero-length strings to be no-ops with no allocations.